File: parser/src/main/java/com/aaroncoplan/waterfall/parser/Pair.java
package com.aaroncoplan.waterfall.parser;

public class Pair<K, V> {

    public final K firstVal;
    public final V secondVal;

    public Pair(K firstVal, V secondVal) {
        this.firstVal = firstVal;
        this.secondVal = secondVal;
    }

    @Override
    public String toString() {
        return "(" + firstVal + ", " + secondVal + ")";
    }
}

File: parser/src/main/kotlin/com/aaroncoplan/waterfall/parser/Pair.kt
package com.aaroncoplan.waterfall.parser

/**
 * A two-element tuple used throughout the codebase (predates Kotlin's `kotlin.Pair`).
 *
 * Java callers reach the fields directly via `pair.firstVal` / `pair.secondVal` thanks to
 * the `@JvmField` annotations; Kotlin callers see them as ordinary properties.
 */
open class Pair<K, V>(
    @JvmField val firstVal: K,
    @JvmField val secondVal: V
) {
    override fun toString(): String = "($firstVal, $secondVal)"
}
